fe/fe-core/src/main/java/org/apache/doris/planner/SingleTabletLoadRecorderMgr.java

```diff
@@ -0,0 +1,119 @@
+                partitionMap.entrySet().removeIf(entry -> entry.getValue().getUpdateTimestamp() < expiryTime);
+                if (partitionMap.isEmpty()) {
+                    dbTableMap.remove(tableId);
+                }
+            });
+            if (dbTableMap.isEmpty()) {
+                loadTabletRecordMap.remove(dbId);
+            }
+        });
+        LOG.info("Remove expired load tablet record successfully.");
+    }
+
+    public int getCurrentLoadTabletIndex(long dbId, long tableId, long partitionId) throws UserException {
+        TabletUpdateRecord record = loadTabletRecordMap.computeIfAbsent(dbId, k -> new ConcurrentHashMap<>())
+                .computeIfAbsent(tableId, k -> new ConcurrentHashMap<>())
+                .get(partitionId);
+        int numBuckets = -1;
+        if (record == null) {
+            numBuckets = getNumBuckets(dbId, tableId, partitionId);
+        }
+        return createOrUpdateLoadTabletRecord(dbId, tableId, partitionId, numBuckets);
+    }
+
+    private int getNumBuckets(long dbId, long tableId, long partitionId) throws UserException {
+        OlapTable olapTable = (OlapTable) Env.getCurrentInternalCatalog().getDb(dbId)
+                .flatMap(db -> db.getTable(tableId)).filter(t -> t.getType() == TableIf.TableType.OLAP)
+                .orElse(null);
+        if (olapTable == null) {
+            throw new UserException("db[" + dbId + "], table[" + tableId + "] is not exist.");
+        }
+        return olapTable.getPartition(partitionId)
+                .getMaterializedIndices(MaterializedIndex.IndexExtState.ALL)
+                .get(0).getTablets().size();
+    }
+
+    private int createOrUpdateLoadTabletRecord(long dbId, long tableId, long partitionId, int numBuckets) {
+        TabletUpdateRecord record = loadTabletRecordMap.computeIfAbsent(dbId, k -> new ConcurrentHashMap<>())
+                .computeIfAbsent(tableId, k -> new ConcurrentHashMap<>())
+                .compute(partitionId, (k, existingRecord) -> {
+                    if (existingRecord == null) {
+                        return new TabletUpdateRecord(0, numBuckets);
+                    } else {
+                        existingRecord.updateRecord();
+                        return existingRecord;
+                    }
+                });
+        return record.getTabletIndex();
+    }
+
+    static class TabletUpdateRecord {
+        @Getter
+        // 0 =< load_tablet_index < number_buckets
+        int tabletIndex;
+        int numBuckets;
+        @Getter
+        long updateTimestamp = System.currentTimeMillis();
+
+        TabletUpdateRecord(int tabletIndex, int numBuckets) {
+            this.tabletIndex = tabletIndex;
+            this.numBuckets = numBuckets;
+        }
+
+        public synchronized void updateRecord() {
+            this.tabletIndex = this.tabletIndex + 1 >= numBuckets ? 0 : this.tabletIndex + 1;
+            this.updateTimestamp = System.currentTimeMillis();
```

Comment:
It's expensive to get system time for each update, just get time if tabletIndex == 0 is better.